### DISCUSSION TITLE
General questions on ArgoCD

### DISCUSSION THREAD

**Discussion by pavan-kumar-ceemala** [argoproj/argo-cd#8164] [General]
Hi,
1st of all, thanks for creating such a wonderful tool, I really liked the ease of deployment, the Git as a source approach, pull mechanism, the sync process etc, this satisfies my small projects, which I've worked till now.
Now I would like to leverage it for a very big project where I need a CD tool which can scale to deploy 500+ microservices per k8s cluster in a multi cluster environment, I've few generic questions to understand if argocd can scale to cater our requirement.

1.  What is an appx upper limit of number of clusters a single argocd can connect and deploy before showing any issues.?
2. How many apps can I create/deploy continuously using argocd before showing any issue?
3. How many max repositories can I configure in argocd?
3. ArgoCD uses kubernetes etcd as its storage, can we externalize this to some other key-value pair storage?

Could you please share any benchmarking reference for the above points, 

Regards,
Pavan

**Comment by meln5674** [comment]
While I'm not at liberty to reveal any specific details, my team has an argocd installation that manages 250+ apps (some of them with close to 100 resources a piece), and while we only have 3 clusters under management, we've had up to 7 previously. We use 2 ui servers, 2 repo servers, and 5 application controllers. While we have had a few performance problems in the past, that was mainly our fault for not setting resource requests and limits on the application controllers, letting them get starved, back when we were still getting to grips with using kubernetes. We're not using particularly beefy hardware either, none of our nodes have more than 16 GiB of memory or 4 cores, as this is our "management" cluster, not one that actually runs "real" apps. As long as you set resource limits and requests to ensure your application controllers and repo servers get the cycles they need, monitor your pod cpu utilization using Prometheus or similar, and scale out as needed, I wouldn't imagine you'd have very many problems. The question isn't "can argocd handle it" but "can argocd scale" and the answer to that is a resounding "yes" in our experience. The only other thing I'd caution is making sure you have enough bandwidth on your management network between argocd and the clusters, as we have found it to be quite "chatty" at times, especially in the situation where a resource gets changed either by a defaulting or mutating webhook, causing it to be permanently out of sync, re-syncing every few minutes.

**Comment by Mais316** [comment]
@meln5674 please advise what was the default resource limits used to have it as a benchmark.
